File: libjava/classpath/gnu/classpath/Pair.java
/* Pair.java -- A heterogenous pair of objects.
   Copyright (C) 2006
   Free Software Foundation, Inc.

This file is part of GNU Classpath.

GNU Classpath is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

GNU Classpath is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with GNU Classpath; see the file COPYING.  If not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA.

Linking this library statically or dynamically with other modules is
making a combined work based on this library.  Thus, the terms and
conditions of the GNU General Public License cover the whole
combination.

As a special exception, the copyright holders of this library give you
permission to link this library with independent modules to produce an
executable, regardless of the license terms of these independent
modules, and to copy and distribute the resulting executable under
terms of your choice, provided that you also meet, for each linked
independent module, the terms and conditions of the license of that
module.  An independent module is a module which is not derived from
or based on this library.  If you modify this library, you may extend
this exception to your version of the library, but you are not
obligated to do so.  If you do not wish to do so, delete this
exception statement from your version. */


package gnu.classpath;

/**
 * A container for a pair of heterogenous objects.
 *
 * @author Andrew John Hughes (gnu_andrew@member.fsf.org)
 */
public class Pair<A,B>
{

  /**
   * The left-hand side of the pair.
   */
  private A left;

  /**
   * The right-hand side of the pair.
   */
  private B right;

  /**
   * Constructs a new pair using the given left and
   * right values.
   *
   * @param left the left-hand side of the pair.
   * @param right the right-hand side of the pair.
   */
  public Pair(A left, B right)
  {
    this.left = left;
    this.right = right;
  }

  /**
   * Returns the left-hand side of the pair.
   *
   * @return the left-hand value.
   */
  public A getLeft()
  {
    return left;
  }

  /**
   * Returns the right-hand side of the pair.
   *
   * @return the right-hand value.
   */
  public B getRight()
  {
    return right;
  }

  /**
   * Returns true if the specified object is also a
   * pair with equivalent left and right values.
   *
   * @param obj the object to compare.
   * @return true if the two are equal.
   */
  public boolean equals(Object obj)
  {
    if (obj instanceof Pair)
      {
        Pair<A,B> p = (Pair<A,B>) obj;
        A lp = p.getLeft();
        B rp = p.getRight();
        return (lp == null ? left == null : lp.equals(left)) &&
          (rp == null ? right == null : rp.equals(right));
      }
    return false;
  }

  /**
   * Returns a hashcode for the pair, created by the
   * summation of the hashcodes of the left and right
   * hand values.
   *
   * @return a hashcode for the pair.
   */
  public int hashCode()
  {
    return (left == null ? 0 : left.hashCode())
      + (right == null ? 0 : right.hashCode());
  }

}

File: libstdc++-v3/testsuite/tr1/6_containers/utility/pair.cc
// 2004-09-23 Chris Jefferson <chris@bubblescope.net>

// Copyright (C) 2004-2019 Free Software Foundation, Inc.
//
// This file is part of the GNU ISO C++ Library.  This library is free
// software; you can redistribute it and/or modify it under the
// terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option)
// any later version.

// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License along
// with this library; see the file COPYING3.  If not see
// <http://www.gnu.org/licenses/>.

// { dg-options "-std=c++98" }

// tr1 additions to pair

#include <tr1/utility>
#include <testsuite_hooks.h>

using namespace std::tr1;
using std::pair;

struct blank_class
{ };

int
main()
{
  typedef pair<int,int> test_pair_type;
  VERIFY(tuple_size<test_pair_type>::value == 2);
  // Test if tuple_element::type returns the correct type
  blank_class blank;
  tuple_element<0, pair<blank_class, int> >::type
    blank2 __attribute__((unused)) = blank;
  tuple_element<1, pair<int ,blank_class> >::type
    blank3 __attribute__((unused)) = blank;
  pair<int,int> test_pair(1, 2);
  VERIFY(get<0>(test_pair) == 1);
  VERIFY(get<1>(test_pair) == 2);
  get<0>(test_pair) = 3;
  get<1>(test_pair) = 4;
  VERIFY(get<0>(test_pair) == 3);
  VERIFY(get<1>(test_pair) == 4);

  const pair<int,int> test_pair2(1,2);
  VERIFY(get<0>(test_pair2) == 1);
  VERIFY(get<1>(test_pair2) == 2);
}

